In [1]:
%load_ext sql

In [2]:
%sql mysql+pymysql://root:root@localhost:3306/md_water_services

Connecting to 'mysql+pymysql://root:***@localhost:3306/md_water_services'

In [3]:
%%sql
DROP TABLE IF EXISTS `auditor_report`;
CREATE TABLE `auditor_report` (
`location_id` VARCHAR(32),
`type_of_water_source` VARCHAR(64),
`true_water_source_score` int DEFAULT NULL,
`statements` VARCHAR(255)
);

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

++
||
++
++

In [5]:
%%sql
SHOW TABLES;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

9 rows affected.

+-----------------------------+
| Tables_in_md_water_services |
+-----------------------------+
|        auditor_report       |
|       data_dictionary       |
|           employee          |
|     global_water_access     |
|           location          |
|            visits           |
|        water_quality        |
|         water_source        |
|        well_pollution       |
+-----------------------------+

In [7]:
%%sql
SELECT 
    * 
FROM 
    auditor_report;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1620 rows affected.

+-------------+----------------------+-------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+
| location_id | type_of_water_source | true_water_source_score |                                                                 statements                                                                |
+-------------+----------------------+-------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+
|  SoRu34980  |         well         |            1            |          Residents admired the official's commitment to enhancing urban life, praising their cooperative and inclusive approach.          |
|  AkRu08112  |         well         |            3            | Villagers spoke highly of the official's dedication and genuine interest in their lives, fostering a sense of belonging and appreciation. |
|  AkLu02044  |        river         |            0            |         Villagers were touched by the official's interactions, noting their humility, strong work ethic, and respectful attitude.         |
|  AkHa00421  |         well         |            3            |   Villagers were moved by the official's visit, praising their hard work, humility, and the profound sense of connection they fostered.   |
|  SoRu35221  |        river         |            0            |                 A photographer's lens captures the queue, though his own struggle for water is a hidden part of the story.                |
|  HaAm16170  |         well         |            1            |               With an open heart, the official created an atmosphere of unity and familial camaraderie among the villagers.               |
|  AkRu04812  |         well         |            3            |     The official's presence left an indelible mark, reflecting their humility, dedication, and the genuine connections they nurtured.     |
|  AkRu08304  |         well         |            3            |         The official's interactions resonated deeply with the villagers, leaving a lasting impression of respect and camaraderie.         |
|  AkRu05107  |         well         |            2            | Villagers spoke highly of the official's dedication and genuine interest in their lives, fostering a sense of belonging and appreciation. |
|  AkRu05215  |         well         |            3            |                  Villagers admired the official's visit for its respectful interactions, hard work, and genuine concern.                  |
+-------------+----------------------+-------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+
Truncated to displaylimit of 10.

## We need to tackle a couple of questions here.
1. Is there a difference in the scores?
2. If so, are there patterns?

 
For the first question, we will have to compare the quality scores in the water_quality table to the auditor's scores. The auditor_report table
used location_id, but the quality scores table only has a record_id we can use. The visits table links location_id and record_id, so we
can link the auditor_report table and water_quality using the visits table.

 

So first, grab the location_id and true_water_source_score columns from auditor_report.

In [9]:
%%sql
SELECT 
    location_id ,
    true_water_source_score
FROM 
    auditor_report;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1620 rows affected.

+-------------+-------------------------+
| location_id | true_water_source_score |
+-------------+-------------------------+
|  SoRu34980  |            1            |
|  AkRu08112  |            3            |
|  AkLu02044  |            0            |
|  AkHa00421  |            3            |
|  SoRu35221  |            0            |
|  HaAm16170  |            1            |
|  AkRu04812  |            3            |
|  AkRu08304  |            3            |
|  AkRu05107  |            2            |
|  AkRu05215  |            3            |
+-------------+-------------------------+
Truncated to displaylimit of 10.

In [13]:
%%sql
DESCRIBE visits;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

7 rows affected.

+----------------------+--------------+------+-----+---------+-------+
|        Field         |     Type     | Null | Key | Default | Extra |
+----------------------+--------------+------+-----+---------+-------+
|      record_id       |     int      |  NO  | PRI |   None  |       |
|     location_id      | varchar(255) | YES  | MUL |   None  |       |
|      source_id       | varchar(510) | YES  | MUL |   None  |       |
|    time_of_record    |   datetime   | YES  |     |   None  |       |
|     visit_count      |     int      | YES  |     |   None  |       |
|    time_in_queue     |     int      | YES  |     |   None  |       |
| assigned_employee_id |     int      | YES  | MUL |   None  |       |
+----------------------+--------------+------+-----+---------+-------+

In [13]:
%%sql
SELECT 
    ar.location_id AS audit_location,
    ar.true_water_source_score,
    vi.location_id AS visit_location,
    vi.record_id
FROM 
    auditor_report ar
INNER JOIN
    visits vi
ON ar.location_id = vi.location_id
LIMIT 10
;


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

10 rows affected.

+----------------+-------------------------+----------------+-----------+
| audit_location | true_water_source_score | visit_location | record_id |
+----------------+-------------------------+----------------+-----------+
|   SoRu34980    |            1            |   SoRu34980    |    5185   |
|   AkRu08112    |            3            |   AkRu08112    |   59367   |
|   AkLu02044    |            0            |   AkLu02044    |   37379   |
|   AkHa00421    |            3            |   AkHa00421    |   51627   |
|   SoRu35221    |            0            |   SoRu35221    |   28758   |
|   HaAm16170    |            1            |   HaAm16170    |   31048   |
|   AkRu04812    |            3            |   AkRu04812    |    1513   |
|   AkRu08304    |            3            |   AkRu08304    |    1218   |
|   AkRu05107    |            2            |   AkRu05107    |    8322   |
|   AkRu05215    |            3            |   AkRu05215    |   21160   |
+----------------+-------------------------+----------------+-----------+
Truncated to displaylimit of 10.

- Now that we have the record_id for each location, our next step is to retrieve the corresponding scores from the water_quality table. We
are particularly interested in the subjective_quality_score. To do this, we'll JOIN the visits table and the water_quality table, using the
record_id as the connecting key.

In [22]:
%%sql

SELECT * FROM water_quality;


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

60146 rows affected.

+-----------+--------------------------+-------------+
| record_id | subjective_quality_score | visit_count |
+-----------+--------------------------+-------------+
|     0     |            0             |      1      |
|     1     |            1             |      1      |
|     2     |            5             |      1      |
|     3     |            10            |      1      |
|     4     |            4             |      1      |
|     5     |            0             |      1      |
|     6     |            9             |      1      |
|     7     |            10            |      1      |
|     8     |            2             |      1      |
|     9     |            10            |      1      |
+-----------+--------------------------+-------------+
Truncated to displaylimit of 10.

In [51]:
%%sql
SELECT 
    ar.location_id AS audit_location,
    ar.true_water_source_score,
    vi.location_id AS visit_location,
    vi.record_id,
    wq.subjective_quality_score
FROM 
    auditor_report ar
INNER JOIN
    visits vi
ON 
    ar.location_id = vi.location_id
INNER JOIN
    water_quality wq

ON 
    vi.record_id = wq.record_id 
    
 ;


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

2698 rows affected.

+----------------+-------------------------+----------------+-----------+--------------------------+
| audit_location | true_water_source_score | visit_location | record_id | subjective_quality_score |
+----------------+-------------------------+----------------+-----------+--------------------------+
|   SoRu34980    |            1            |   SoRu34980    |    5185   |            1             |
|   AkRu08112    |            3            |   AkRu08112    |   59367   |            3             |
|   AkLu02044    |            0            |   AkLu02044    |   37379   |            0             |
|   AkHa00421    |            3            |   AkHa00421    |   51627   |            3             |
|   SoRu35221    |            0            |   SoRu35221    |   28758   |            0             |
|   HaAm16170    |            1            |   HaAm16170    |   31048   |            1             |
|   AkRu04812    |            3            |   AkRu04812    |    1513   |            3             |
|   AkRu08304    |            3            |   AkRu08304    |    1218   |            3             |
|   AkRu05107    |            2            |   AkRu05107    |    8322   |            2             |
|   AkRu05215    |            3            |   AkRu05215    |   21160   |            10            |
+----------------+-------------------------+----------------+-----------+--------------------------+
Truncated to displaylimit of 10.

- It doesn't matter if your columns are in a different format, because we are about to clean this up a bit. Since it is a duplicate, we can drop one of
the location_id columns. Let's leave record_id and rename the scores to surveyor_score and auditor_score to make it clear which scores
we're looking at in the results set.

In [39]:
%%sql
SELECT 
    vi.location_id AS visit_location,
    vi.record_id,
    ar.true_water_source_score AS auditor_score,
    wq.subjective_quality_score AS surveyor_score
FROM 
    auditor_report ar
INNER JOIN
    visits vi
ON 
    ar.location_id = vi.location_id
INNER JOIN
    water_quality wq

ON 
    vi.record_id = wq.record_id 
    
 
;


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

2698 rows affected.

+----------------+-----------+---------------+----------------+
| visit_location | record_id | auditor_score | surveyor_score |
+----------------+-----------+---------------+----------------+
|   SoRu34980    |    5185   |       1       |       1        |
|   AkRu08112    |   59367   |       3       |       3        |
|   AkLu02044    |   37379   |       0       |       0        |
|   AkHa00421    |   51627   |       3       |       3        |
|   SoRu35221    |   28758   |       0       |       0        |
|   HaAm16170    |   31048   |       1       |       1        |
|   AkRu04812    |    1513   |       3       |       3        |
|   AkRu08304    |    1218   |       3       |       3        |
|   AkRu05107    |    8322   |       2       |       2        |
|   AkRu05215    |   21160   |       3       |       10       |
+----------------+-----------+---------------+----------------+
Truncated to displaylimit of 10.

- Since were joining 1620 rows of data, we want to keep track of the number of rows we get each time we run a query. We can either set the
maximum number of rows we want from "Limit to 1000 rows" to a larger number like 10000, or we can force SQL to give us all of the results, using
LIMIT 10000.
- Ok, let's analyse! A good starting point is to check if the auditor's and exployees' scores agree. There are many ways to do it. We can have a
WHERE clause and check if surveyor_score = auditor_score, or we can subtract the two scores and check if the result is 0.

In [49]:
%%sql
SELECT 
    vi.location_id AS visit_location,
    vi.record_id,
    ar.true_water_source_score AS auditor_score,
    wq.subjective_quality_score AS surveyor_score
FROM 
    auditor_report ar

INNER JOIN
    visits vi
ON 
    ar.location_id = vi.location_id
INNER JOIN
    water_quality wq

ON 
    vi.record_id = wq.record_id 
WHERE
    ar.true_water_source_score = wq.subjective_quality_score   
 
;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

2505 rows affected.

+----------------+-----------+---------------+----------------+
| visit_location | record_id | auditor_score | surveyor_score |
+----------------+-----------+---------------+----------------+
|   SoRu34980    |    5185   |       1       |       1        |
|   AkRu08112    |   59367   |       3       |       3        |
|   AkLu02044    |   37379   |       0       |       0        |
|   AkHa00421    |   51627   |       3       |       3        |
|   SoRu35221    |   28758   |       0       |       0        |
|   HaAm16170    |   31048   |       1       |       1        |
|   AkRu04812    |    1513   |       3       |       3        |
|   AkRu08304    |    1218   |       3       |       3        |
|   AkRu05107    |    8322   |       2       |       2        |
|   HaDe16541    |   13070   |       2       |       2        |
+----------------+-----------+---------------+----------------+
Truncated to displaylimit of 10.

- You try and figure this one out. If you run into an error, remember that you can't use aliases in WHERE, so you have to use the same name as in the
SELECT part, like this: auditor_report.true_water_source_score

 
- You got 2505 rows right? Some of the locations were visited multiple times, so these records are duplicated here. To fix it, we set visits.visit_count
= 1 in the WHERE clause. Make sure you reference the alias you used for visits in the join.

 

- With the duplicates removed I now get 1518. What does this mean considering the auditor visited 1620 sites?

In [55]:
%%sql
SELECT * FROM visits;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

60146 rows affected.

+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
| record_id | location_id |  source_id   |    time_of_record   | visit_count | time_in_queue | assigned_employee_id |
+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
|     0     |  SoIl32582  | SoIl32582224 | 2021-01-01 09:10:00 |      1      |       15      |          12          |
|     1     |  KiRu28935  | KiRu28935224 | 2021-01-01 09:17:00 |      1      |       0       |          46          |
|     2     |  HaRu19752  | HaRu19752224 | 2021-01-01 09:36:00 |      1      |       62      |          40          |
|     3     |  AkLu01628  | AkLu01628224 | 2021-01-01 09:53:00 |      1      |       0       |          1           |
|     4     |  AkRu03357  | AkRu03357224 | 2021-01-01 10:11:00 |      1      |       28      |          14          |
|     5     |  KiRu29315  | KiRu29315224 | 2021-01-01 10:17:00 |      1      |       9       |          40          |
|     6     |  AkRu05234  | AkRu05234224 | 2021-01-01 10:18:00 |      1      |       0       |          30          |
|     7     |  KiRu28520  | KiRu28520224 | 2021-01-01 10:28:00 |      1      |       0       |          34          |
|     8     |  HaZa21742  | HaZa21742224 | 2021-01-01 10:37:00 |      1      |       0       |          6           |
|     9     |  AmDa12214  | AmDa12214224 | 2021-01-01 10:58:00 |      1      |       0       |          36          |
+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
Truncated to displaylimit of 10.

In [58]:
%%sql
SELECT 
    vi.location_id AS visit_location,
    vi.record_id,
    ar.true_water_source_score AS auditor_score,
    wq.subjective_quality_score AS surveyor_score
FROM 
    auditor_report ar

INNER JOIN
    visits vi
ON 
    ar.location_id = vi.location_id
AND 
        vi.visit_count = 1
INNER JOIN
    water_quality wq

ON 
    vi.record_id = wq.record_id 
WHERE
    ar.true_water_source_score = wq.subjective_quality_score   
 
;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1518 rows affected.

+----------------+-----------+---------------+----------------+
| visit_location | record_id | auditor_score | surveyor_score |
+----------------+-----------+---------------+----------------+
|   SoRu34980    |    5185   |       1       |       1        |
|   AkRu08112    |   59367   |       3       |       3        |
|   AkLu02044    |   37379   |       0       |       0        |
|   AkHa00421    |   51627   |       3       |       3        |
|   SoRu35221    |   28758   |       0       |       0        |
|   HaAm16170    |   31048   |       1       |       1        |
|   AkRu04812    |    1513   |       3       |       3        |
|   AkRu08304    |    1218   |       3       |       3        |
|   AkRu05107    |    8322   |       2       |       2        |
|   HaDe16541    |   13070   |       2       |       2        |
+----------------+-----------+---------------+----------------+
Truncated to displaylimit of 10.

- I think that is an excellent result. 1518/1620 = 94% of the records the auditor checked were correct!!

 

- But that means that 102 records are incorrect. So let's look at those. You can do it by adding one character in the last query!

In [60]:
%%sql
SELECT 
    vi.location_id AS visit_location,
    vi.record_id,
    ar.true_water_source_score AS auditor_score,
    wq.subjective_quality_score AS surveyor_score
FROM 
    auditor_report ar
INNER JOIN
    visits vi
ON 
    ar.location_id = vi.location_id
AND
    vi.visit_count = 1
LEFT JOIN
    water_quality wq
ON 
    vi.record_id = wq.record_id 
WHERE
    ar.true_water_source_score != wq.subjective_quality_score
OR
    wq.subjective_quality_score IS NULL;


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

102 rows affected.

+----------------+-----------+---------------+----------------+
| visit_location | record_id | auditor_score | surveyor_score |
+----------------+-----------+---------------+----------------+
|   AkRu05215    |   21160   |       3       |       10       |
|   KiRu29290    |    7938   |       3       |       10       |
|   KiHa22748    |   43140   |       9       |       10       |
|   SoRu37841    |   18495   |       6       |       10       |
|   KiRu27884    |   33931   |       1       |       10       |
|   KiZu31170    |   17950   |       9       |       10       |
|   KiZu31370    |   36864   |       3       |       10       |
|   AkRu06495    |   45924   |       2       |       10       |
|   HaRu17528    |   30524   |       1       |       10       |
|   SoRu38331    |   13192   |       3       |       10       |
+----------------+-----------+---------------+----------------+
Truncated to displaylimit of 10.

In [67]:
%%sql
SELECT * FROM water_source;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

39650 rows affected.

+--------------+----------------------+-------------------------+
|  source_id   | type_of_water_source | number_of_people_served |
+--------------+----------------------+-------------------------+
| AkHa00000224 |     tap_in_home      |           956           |
| AkHa00001224 |  tap_in_home_broken  |           930           |
| AkHa00002224 |  tap_in_home_broken  |           486           |
| AkHa00003224 |         well         |           364           |
| AkHa00004224 |  tap_in_home_broken  |           942           |
| AkHa00005224 |     tap_in_home      |           736           |
| AkHa00006224 |     tap_in_home      |           882           |
| AkHa00007224 |     tap_in_home      |           554           |
| AkHa00008224 |         well         |           398           |
| AkHa00009224 |         well         |           346           |
+--------------+----------------------+-------------------------+
Truncated to displaylimit of 10.

- Since we used some of this data in our previous analyses, we need to make sure those results are still valid, now we know some of them are
incorrect. We didn't use the scores that much, but we relied a lot on the type_of_water_source, so let's check if there are any errors there.
 

- So, to do this, we need to grab the type_of_water_source column from the water_source table and call it survey_source, using the
source_id column to JOIN. Also select the type_of_water_source from the auditor_report table, and call it auditor_source.

In [83]:
%%sql
SELECT 
    ar.location_id AS visit_location,
    vi.record_id,
    ar.type_of_water_source AS auditor_source,
    ws.type_of_water_source AS survey_source,
    ar.true_water_source_score AS auditor_score,
    wq.subjective_quality_score AS surveyor_score
FROM 
    auditor_report ar
INNER JOIN
    visits vi
ON 
    ar.location_id = vi.location_id
AND
    vi.visit_count = 1
INNER JOIN
    water_quality wq
ON 
    vi.record_id = wq.record_id
INNER JOIN
    water_source ws
ON 
    vi.source_id = ws.source_id
WHERE
    ar.true_water_source_score != wq.subjective_quality_score
OR
    wq.subjective_quality_score IS NULL
OR
    ar.type_of_water_source != ws.type_of_water_source
;


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

102 rows affected.

+----------------+-----------+--------------------+--------------------+---------------+----------------+
| visit_location | record_id |   auditor_source   |   survey_source    | auditor_score | surveyor_score |
+----------------+-----------+--------------------+--------------------+---------------+----------------+
|   AkRu05215    |   21160   |        well        |        well        |       3       |       10       |
|   KiRu29290    |    7938   |     shared_tap     |     shared_tap     |       3       |       10       |
|   KiHa22748    |   43140   | tap_in_home_broken | tap_in_home_broken |       9       |       10       |
|   SoRu37841    |   18495   |     shared_tap     |     shared_tap     |       6       |       10       |
|   KiRu27884    |   33931   |        well        |        well        |       1       |       10       |
|   KiZu31170    |   17950   | tap_in_home_broken | tap_in_home_broken |       9       |       10       |
|   KiZu31370    |   36864   |     shared_tap     |     shared_tap     |       3       |       10       |
|   AkRu06495    |   45924   |        well        |        well        |       2       |       10       |
|   HaRu17528    |   30524   |        well        |        well        |       1       |       10       |
|   SoRu38331    |   13192   |     shared_tap     |     shared_tap     |       3       |       10       |
+----------------+-----------+--------------------+--------------------+---------------+----------------+
Truncated to displaylimit of 10.

## Linking records to employees
Next up, let's look at where these errors may have come from. At some of the locations, employees assigned scores incorrectly, and those records
ended up in this results set.

 

### I think there are two reasons this can happen.
1. These workers are all humans and make mistakes so this is expected.
2. Unfortunately, the alternative is that someone assigned scores incorrectly on purpose!

 
In either case, the employees are the source of the errors, so let's JOIN the assigned_employee_id for all the people on our list from the visits
table to our query. Remember, our query shows the shows the 102 incorrect records, so when we join the employee data, we can see which
employees made these incorrect records.

In [85]:
%%sql
SELECT 
    vi.location_id AS visit_location,
    vi.record_id,
    ar.true_water_source_score AS auditor_score,
    wq.subjective_quality_score AS surveyor_score,
    vi.assigned_employee_id
FROM 
    auditor_report ar
INNER JOIN
    visits vi
ON 
    ar.location_id = vi.location_id
AND
    vi.visit_count = 1
LEFT JOIN
    water_quality wq
ON 
    vi.record_id = wq.record_id 
WHERE
    ar.true_water_source_score != wq.subjective_quality_score
OR
    wq.subjective_quality_score IS NULL;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

102 rows affected.

+----------------+-----------+---------------+----------------+----------------------+
| visit_location | record_id | auditor_score | surveyor_score | assigned_employee_id |
+----------------+-----------+---------------+----------------+----------------------+
|   AkRu05215    |   21160   |       3       |       10       |          34          |
|   KiRu29290    |    7938   |       3       |       10       |          1           |
|   KiHa22748    |   43140   |       9       |       10       |          1           |
|   SoRu37841    |   18495   |       6       |       10       |          34          |
|   KiRu27884    |   33931   |       1       |       10       |          1           |
|   KiZu31170    |   17950   |       9       |       10       |          5           |
|   KiZu31370    |   36864   |       3       |       10       |          48          |
|   AkRu06495    |   45924   |       2       |       10       |          1           |
|   HaRu17528    |   30524   |       1       |       10       |          18          |
|   SoRu38331    |   13192   |       3       |       10       |          5           |
+----------------+-----------+---------------+----------------+----------------------+
Truncated to displaylimit of 10.

- So now we can link the incorrect records to the employees who recorded them. The ID's don't help us to identify them. We have employees' names
stored along with their IDs, so let's fetch their names from the employees table instead of the ID's.

In [88]:
%%sql
SELECT * FROM employee;


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

56 rows affected.

+----------------------+----------------+--------------+-------------------------------+---------------------------+---------------+-----------+----------------+
| assigned_employee_id | employee_name  | phone_number |             email             |          address          | province_name | town_name |    position    |
+----------------------+----------------+--------------+-------------------------------+---------------------------+---------------+-----------+----------------+
|          0           |  Amara Jengo   | +99637993287 |   amara.jengo@ndogowater.gov  |  36 Pwani Mchangani Road  |     Sokoto    |   Ilanga  | Field Surveyor |
|          1           |  Bello Azibo   | +99643864786 |   bello.azibo@ndogowater.gov  |   129 Ziwa La Kioo Road   |    Kilimani   |   Rural   | Field Surveyor |
|          2           |  Bakari Iniko  | +99222599041 |  bakari.iniko@ndogowater.gov  |   18 Mlima Tazama Avenue  |    Hawassa    |   Rural   | Field Surveyor |
|          3           | Malachi Mavuso | +99945849900 | malachi.mavuso@ndogowater.gov |     100 Mogadishu Road    |     Akatsi    |   Lusaka  | Field Surveyor |
|          4           |  Cheche Buhle  | +99381679640 |  cheche.buhle@ndogowater.gov  |      1 Savanna Street     |     Akatsi    |   Rural   | Field Surveyor |
|          5           | Zuriel Matembo | +99034075111 | zuriel.matembo@ndogowater.gov |  26 Bahari Ya Faraja Road |    Kilimani   |   Rural   | Field Surveyor |
|          6           |  Deka Osumare  | +99379364631 |  deka.osumare@ndogowater.gov  |    104 Kenyatta Street    |     Akatsi    |   Rural   | Field Surveyor |
|          7           | Lalitha Kaburi | +99681623240 | lalitha.kaburi@ndogowater.gov | 145 Sungura Amanpour Road |    Kilimani   |   Rural   | Field Surveyor |
|          8           |  Enitan Zuri   | +99248509202 |   enitan.zuri@ndogowater.gov  |      117 Kampala Road     |    Hawassa    |  Zanzibar | Field Surveyor |
|          10          |   Farai Nia    | +99570082739 |    farai.nia@ndogowater.gov   | 33 Angélique Kidjo Avenue |     Amanzi    |   Dahabu  | Field Surveyor |
+----------------------+----------------+--------------+-------------------------------+---------------------------+---------------+-----------+----------------+
Truncated to displaylimit of 10.

In [90]:
%%sql
SELECT 
    vi.location_id AS visit_location,
    vi.record_id,
    ar.true_water_source_score AS auditor_score,
    wq.subjective_quality_score AS surveyor_score,
    vi.assigned_employee_id,
    e.employee_name
FROM 
    auditor_report ar
INNER JOIN
    visits vi
ON 
    ar.location_id = vi.location_id
AND
    vi.visit_count = 1
LEFT JOIN
    water_quality wq
ON 
    vi.record_id = wq.record_id 
INNER JOIN
    employee e
ON 
    vi.assigned_employee_id = e.assigned_employee_id
WHERE
    ar.true_water_source_score != wq.subjective_quality_score
OR
    wq.subjective_quality_score IS NULL;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

102 rows affected.

+----------------+-----------+---------------+----------------+----------------------+----------------+
| visit_location | record_id | auditor_score | surveyor_score | assigned_employee_id | employee_name  |
+----------------+-----------+---------------+----------------+----------------------+----------------+
|   AkRu05215    |   21160   |       3       |       10       |          34          |   Rudo Imani   |
|   KiRu29290    |    7938   |       3       |       10       |          1           |  Bello Azibo   |
|   KiHa22748    |   43140   |       9       |       10       |          1           |  Bello Azibo   |
|   SoRu37841    |   18495   |       6       |       10       |          34          |   Rudo Imani   |
|   KiRu27884    |   33931   |       1       |       10       |          1           |  Bello Azibo   |
|   KiZu31170    |   17950   |       9       |       10       |          5           | Zuriel Matembo |
|   KiZu31370    |   36864   |       3       |       10       |          48          | Yewande Ebele  |
|   AkRu06495    |   45924   |       2       |       10       |          1           |  Bello Azibo   |
|   HaRu17528    |   30524   |       1       |       10       |          18          | Jengo Tumaini  |
|   SoRu38331    |   13192   |       3       |       10       |          5           | Zuriel Matembo |
+----------------+-----------+---------------+----------------+----------------------+----------------+
Truncated to displaylimit of 10.

- Let's first get a unique list of employees from this table. Think back to the start of your SQL journey to answer this one. I got 17 employees.

In [98]:
%%sql
WITH Incorrect_records AS (
    SELECT 
    vi.location_id AS visit_location,
    vi.record_id,
    ar.true_water_source_score AS auditor_score,
    wq.subjective_quality_score AS surveyor_score,
    vi.assigned_employee_id,
    e.employee_name
FROM 
    auditor_report ar
INNER JOIN
    visits vi
ON 
    ar.location_id = vi.location_id
AND
    vi.visit_count = 1
LEFT JOIN
    water_quality wq
ON 
    vi.record_id = wq.record_id 
INNER JOIN
    employee e
ON 
    vi.assigned_employee_id = e.assigned_employee_id
WHERE
    ar.true_water_source_score != wq.subjective_quality_score
OR
    wq.subjective_quality_score IS NULL

)

SELECT DISTINCT
    employee_name 
FROM
    Incorrect_records;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

17 rows affected.

+----------------+
| employee_name  |
+----------------+
|   Rudo Imani   |
|  Bello Azibo   |
| Zuriel Matembo |
| Yewande Ebele  |
| Jengo Tumaini  |
|   Farai Nia    |
| Malachi Mavuso |
|  Makena Thabo  |
| Lalitha Kaburi |
|  Gamba Shani   |
+----------------+
Truncated to displaylimit of 10.

- Next, let's try to calculate how many mistakes each employee made. So basically we want to count how many times their name is in
Incorrect_records list, and then group them by name, right?

In [101]:
%%sql
WITH Incorrect_records AS (
    SELECT 
    vi.location_id AS visit_location,
    vi.record_id,
    ar.true_water_source_score AS auditor_score,
    wq.subjective_quality_score AS surveyor_score,
    vi.assigned_employee_id,
    e.employee_name
FROM 
    auditor_report ar
INNER JOIN
    visits vi
ON 
    ar.location_id = vi.location_id
AND
    vi.visit_count = 1
LEFT JOIN
    water_quality wq
ON 
    vi.record_id = wq.record_id 
INNER JOIN
    employee e
ON 
    vi.assigned_employee_id = e.assigned_employee_id
WHERE
    ar.true_water_source_score != wq.subjective_quality_score
OR
    wq.subjective_quality_score IS NULL

)

SELECT 
    employee_name,
    COUNT(*) AS mistake_count
FROM 
    Incorrect_records
GROUP BY 
    employee_name
ORDER BY 
    mistake_count DESC;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

17 rows affected.

+----------------+---------------+
| employee_name  | mistake_count |
+----------------+---------------+
|  Bello Azibo   |       26      |
| Malachi Mavuso |       21      |
| Zuriel Matembo |       17      |
| Lalitha Kaburi |       7       |
|   Rudo Imani   |       5       |
|   Farai Nia    |       4       |
|  Enitan Zuri   |       4       |
| Yewande Ebele  |       3       |
| Jengo Tumaini  |       3       |
|  Makena Thabo  |       3       |
+----------------+---------------+
Truncated to displaylimit of 10.

## Gathering some evidence
Ok, so thinking about this a bit. How would we go about finding out if any of our employees are corrupt?

 
- Let's say all employees make mistakes, if someone is corrupt, they will be making a lot of "mistakes", more than average, for example. But someone
could just be clumsy, so we should try to get more evidence...

 
- Our auditor did say some of the things he heard on the streets were quite shady, and he recorded this in the statements column. Considering
both of these sources should give us a pretty reliable answer.

### 2. So let's try to find all of the employees who have an above-average number of mistakes. Let's break it down into steps first:
1. We have to first calculate the number of times someone's name comes up. (we just did that in the previous query). Let's call it error_count.
2. Then, we need to calculate the average number of mistakes employees made. We can do that by taking the average of the previous query's
results. Something like this:
SELECT
AVG(number_of_mistakes)
FROM
error_count;
Let's call that result avg_error_count_per_empl, which would be a scalar value.


In [117]:
%%sql
WITH Incorrect_records AS (
    SELECT 
    vi.location_id AS visit_location,
    vi.record_id,
    ar.true_water_source_score AS auditor_score,
    wq.subjective_quality_score AS surveyor_score,
    vi.assigned_employee_id,
    e.employee_name
FROM 
    auditor_report ar
INNER JOIN
    visits vi
ON 
    ar.location_id = vi.location_id
AND
    vi.visit_count = 1
LEFT JOIN
    water_quality wq
ON 
    vi.record_id = wq.record_id 
INNER JOIN
    employee e
ON 
    vi.assigned_employee_id = e.assigned_employee_id
WHERE
    ar.true_water_source_score != wq.subjective_quality_score
OR
    wq.subjective_quality_score IS NULL

),

error_count AS (
SELECT 
    employee_name,
    COUNT(*) AS number_of_mistake
FROM 
    Incorrect_records
GROUP BY 
    employee_name

)
 
SELECT
    employee_name, 
    AVG(number_of_mistake)
FROM
    error_count
GROUP BY 
    employee_name




Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

17 rows affected.

+----------------+------------------------+
| employee_name  | AVG(number_of_mistake) |
+----------------+------------------------+
|   Rudo Imani   |         5.0000         |
|  Bello Azibo   |        26.0000         |
| Zuriel Matembo |        17.0000         |
| Yewande Ebele  |         3.0000         |
| Jengo Tumaini  |         3.0000         |
|   Farai Nia    |         4.0000         |
| Malachi Mavuso |        21.0000         |
|  Makena Thabo  |         3.0000         |
| Lalitha Kaburi |         7.0000         |
|  Gamba Shani   |         3.0000         |
+----------------+------------------------+
Truncated to displaylimit of 10.

### 3. Finaly we have to compare each employee's error_count with avg_error_count_per_empl. We will call this results set our suspect_list.
- Remember that we can't use an aggregate result in WHERE, so we have to use avg_error_count_per_empl as a subquery.
SELECT
employee_name,
number_of_mistakes
FROM
error_count
WHERE
number_of_mistakes > (avg_error_count_per_empl);

In [136]:
%%sql
WITH Incorrect_records AS (
    SELECT 
    vi.location_id AS visit_location,
    vi.record_id,
    ar.true_water_source_score AS auditor_score,
    wq.subjective_quality_score AS surveyor_score,
    vi.assigned_employee_id,
    e.employee_name
FROM 
    auditor_report ar
INNER JOIN
    visits vi
ON 
    ar.location_id = vi.location_id
AND
    vi.visit_count = 1
LEFT JOIN
    water_quality wq
ON 
    vi.record_id = wq.record_id 
INNER JOIN
    employee e
ON 
    vi.assigned_employee_id = e.assigned_employee_id
WHERE
    ar.true_water_source_score != wq.subjective_quality_score
OR
    wq.subjective_quality_score IS NULL

),

error_count AS (
SELECT 
    employee_name,
    COUNT(*) AS number_of_mistake
FROM 
    Incorrect_records
GROUP BY 
    employee_name

),

Average_Error AS  (
SELECT
     employee_name, 
    AVG(number_of_mistake)
FROM
    error_count
GROUP BY 
    employee_name

)
SELECT
    ec.employee_name,
    ec.number_of_mistakes
FROM
    error_count ec
CROSS JOIN
    Average_Error ae
WHERE
    ec.number_of_mistakes > ae.avg_error_count_per_empl;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

RuntimeError: (pymysql.err.OperationalError) (1054, "Unknown column 'ec.number_of_mistakes' in 'field list'")
[SQL: WITH Incorrect_records AS (
    SELECT
    vi.location_id AS visit_location,
    vi.record_id,
    ar.true_water_source_score AS auditor_score,
    wq.subjective_quality_score AS surveyor_score,
    vi.assigned_employee_id,
    e.employee_name
FROM
    auditor_report ar
INNER JOIN
    visits vi
ON
    ar.location_id = vi.location_id
AND
    vi.visit_count = 1
LEFT JOIN
    water_quality wq
ON
    vi.record_id = wq.record_id
INNER JOIN
    employee e
ON
    vi.assigned_employee_id = e.assigned_employee_id
WHERE
    ar.true_water_source_score != wq.subjective_quality_score
OR
    wq.subjective_quality_score IS NULL

),

error_count AS (
SELECT
    employee_name,
    COUNT(*) AS number_of_mistake
FROM
    Incorrect_records
GROUP BY
    employee_name

),

Average_Error AS  (
SELECT
     employee_name,
    AVG(number_of_mistake)
FROM
    error_count
GROUP BY
    employee_name


- View table

In [141]:
%%sql
CREATE VIEW  Incorrect_records AS (
SELECT 
    vi.location_id AS visit_location,
    vi.record_id,
    ar.true_water_source_score AS auditor_score,
    wq.subjective_quality_score AS surveyor_score,
    vi.assigned_employee_id,
    e.employee_name
FROM 
    auditor_report ar
INNER JOIN
    visits vi
ON 
    ar.location_id = vi.location_id
AND
    vi.visit_count = 1
JOIN
    water_quality wq
ON 
    vi.record_id = wq.record_id 
INNER JOIN
    employee e
ON 
    vi.assigned_employee_id = e.assigned_employee_id
WHERE
    vi.visit_count = 1 AND
    ar.true_water_source_score != wq.subjective_quality_score
 
);

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

++
||
++
++

In [145]:
%%sql
SELECT * FROM Incorrect_records;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

102 rows affected.

+----------------+-----------+---------------+----------------+----------------------+----------------+
| visit_location | record_id | auditor_score | surveyor_score | assigned_employee_id | employee_name  |
+----------------+-----------+---------------+----------------+----------------------+----------------+
|   AkRu05215    |   21160   |       3       |       10       |          34          |   Rudo Imani   |
|   KiRu29290    |    7938   |       3       |       10       |          1           |  Bello Azibo   |
|   KiHa22748    |   43140   |       9       |       10       |          1           |  Bello Azibo   |
|   SoRu37841    |   18495   |       6       |       10       |          34          |   Rudo Imani   |
|   KiRu27884    |   33931   |       1       |       10       |          1           |  Bello Azibo   |
|   KiZu31170    |   17950   |       9       |       10       |          5           | Zuriel Matembo |
|   KiZu31370    |   36864   |       3       |       10       |          48          | Yewande Ebele  |
|   AkRu06495    |   45924   |       2       |       10       |          1           |  Bello Azibo   |
|   HaRu17528    |   30524   |       1       |       10       |          18          | Jengo Tumaini  |
|   SoRu38331    |   13192   |       3       |       10       |          5           | Zuriel Matembo |
+----------------+-----------+---------------+----------------+----------------------+----------------+
Truncated to displaylimit of 10.

In [153]:
%%sql

WITH error_count AS (  
SELECT
    employee_name,
    COUNT(employee_name) AS number_of_mistakes
FROM
    Incorrect_records
 
GROUP BY
    employee_name
ORDER BY
    number_of_mistakes DESC
)
 
SELECT * FROM error_count;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

17 rows affected.

+----------------+--------------------+
| employee_name  | number_of_mistakes |
+----------------+--------------------+
|  Bello Azibo   |         26         |
| Malachi Mavuso |         21         |
| Zuriel Matembo |         17         |
| Lalitha Kaburi |         7          |
|   Rudo Imani   |         5          |
|   Farai Nia    |         4          |
|  Enitan Zuri   |         4          |
| Yewande Ebele  |         3          |
| Jengo Tumaini  |         3          |
|  Makena Thabo  |         3          |
+----------------+--------------------+
Truncated to displaylimit of 10.

 ## 2. Now calculate the average of the number_of_mistakes in error_count. You should get a single value.
 


In [162]:
%%sql
WITH error_count AS (  
SELECT
    employee_name,
    COUNT(employee_name) AS number_of_mistakes
FROM
    Incorrect_records
 
GROUP BY
    employee_name
ORDER BY
    number_of_mistakes DESC
)
 
SELECT AVG( number_of_mistakes) FROM error_count;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

+--------------------------+
| AVG( number_of_mistakes) |
+--------------------------+
|          6.0000          |
+--------------------------+

## 3.To find the employees who made more mistakes than the average person, we need the employee's names, the number of mistakes each one made, and filter the employees with an above-average number of mistakes.
###### HINT: Use SELECT AVG(mistake_count) FROM error_count as a custom filter in the WHERE part of our query.

In [237]:
%%sql

WITH error_count AS (
    SELECT
        employee_name,
        COUNT(employee_name) AS number_of_mistakes
    FROM
        Incorrect_records
    GROUP BY
        employee_name
    ORDER BY
        number_of_mistakes DESC
),
average_mistakes AS (
    SELECT AVG(number_of_mistakes) AS avg_mistakes
    FROM error_count
)
SELECT
    ec.employee_name,
    ec.number_of_mistakes
FROM
    error_count ec
JOIN
    average_mistakes am
ON
    ec.number_of_mistakes > am.avg_mistakes
ORDER BY
    ec.number_of_mistakes DESC;



Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

4 rows affected.

+----------------+--------------------+
| employee_name  | number_of_mistakes |
+----------------+--------------------+
|  Bello Azibo   |         26         |
| Malachi Mavuso |         21         |
| Zuriel Matembo |         17         |
| Lalitha Kaburi |         7          |
+----------------+--------------------+

- We should look at the Incorrect_records table again, and isolate all of the records these four employees gathered. We should also look at the
statements for these records to look for patterns.

 
- First, convert the suspect_list to a CTE, so we can use it to filter the records from these four employees. Make sure you get the names of the
four "suspects", and their mistake count as a result, using SELECT employee_name FROM suspect_list.

In [177]:
%%sql
WITH error_count AS (
    SELECT
        employee_name,
        COUNT(employee_name) AS number_of_mistakes
    FROM
        Incorrect_records
    GROUP BY
        employee_name
),
suspect_list AS (
    SELECT
        employee_name
    FROM
        error_count
    WHERE
        employee_name IN ('Bello Azibo', 'Malachi Mavuso', 'Zuriel Matembo', 'Lalitha Kaburi')  
)

-- Isolate records gathered by the suspect employees
SELECT
    ir.visit_location,
    ir.record_id,
    ir.auditor_score,
    ir.surveyor_score,
    ir.assigned_employee_id,
    ir.employee_name
FROM
    Incorrect_records ir
JOIN
    suspect_list sl
ON
    ir.employee_name = sl.employee_name;


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

71 rows affected.

+----------------+-----------+---------------+----------------+----------------------+---------------+
| visit_location | record_id | auditor_score | surveyor_score | assigned_employee_id | employee_name |
+----------------+-----------+---------------+----------------+----------------------+---------------+
|   AmRu14842    |   47055   |       2       |       10       |          1           |  Bello Azibo  |
|   HaYa21395    |   28729   |       2       |       10       |          1           |  Bello Azibo  |
|   AkRu05741    |   19961   |       3       |       10       |          1           |  Bello Azibo  |
|   AmRu14887    |   31028   |       9       |       10       |          1           |  Bello Azibo  |
|   KiMr24919    |   23328   |       1       |       10       |          1           |  Bello Azibo  |
|   KiRu29639    |   45008   |       1       |       10       |          1           |  Bello Azibo  |
|   KiAm22092    |    8186   |       9       |       10       |          1           |  Bello Azibo  |
|   AkLu02500    |   24797   |       3       |       10       |          1           |  Bello Azibo  |
|   AkRu03848    |   43412   |       3       |       10       |          1           |  Bello Azibo  |
|   AkRu04508    |   37357   |       1       |       10       |          1           |  Bello Azibo  |
+----------------+-----------+---------------+----------------+----------------------+---------------+
Truncated to displaylimit of 10.

## You should get a column of names back. So let's just recap here...
1. We use Incorrect_records to find all of the records where the auditor and employee scores don't match.
2. We then used error_count to aggregate the data, and got the number of mistakes each employee made.
3. Finally, suspect_list retrieves the data of employees who make an above-average number of mistakes.
Now we can filter that Incorrect_records view to identify all of the records associated with the four employees we identified.

In [182]:
%%sql
WITH error_count AS (  
    SELECT
        employee_name,
        COUNT(*) AS number_of_mistakes
    FROM
        Incorrect_records
    GROUP BY
        employee_name
    ORDER BY
        number_of_mistakes DESC
)

-- Retrieve the aggregated data
SELECT
    employee_name,
    number_of_mistakes
FROM
    error_count;


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

17 rows affected.

+----------------+--------------------+
| employee_name  | number_of_mistakes |
+----------------+--------------------+
|  Bello Azibo   |         26         |
| Malachi Mavuso |         21         |
| Zuriel Matembo |         17         |
| Lalitha Kaburi |         7          |
|   Rudo Imani   |         5          |
|   Farai Nia    |         4          |
|  Enitan Zuri   |         4          |
| Yewande Ebele  |         3          |
| Jengo Tumaini  |         3          |
|  Makena Thabo  |         3          |
+----------------+--------------------+
Truncated to displaylimit of 10.

In [192]:
%%sql
-- Step 1: Count mistakes per employee
WITH error_count AS (  
    SELECT
        employee_name,
        COUNT(*) AS number_of_mistakes
    FROM
        Incorrect_records
    GROUP BY
        employee_name
    ORDER BY
        number_of_mistakes DESC
),

-- Step 2: Calculate the average number of mistakes
    
average_mistakes AS( 
    SELECT 
        AVG(number_of_mistakes) AS avg_mistakes
    FROM 
        error_count
),
-- Step 3: Identify employees with more mistakes than the average
    
suspect_list AS (
    SELECT
        ec.employee_name
    FROM
        error_count ec
    JOIN
        average_mistakes am
    ON
        ec.number_of_mistakes > am.avg_mistakes
)
-- Step 4: Filter Incorrect_records to find records associated with suspect employees
-- Retrieve records from Incorrect_records where employee_name is in the suspect_list
SELECT
    ir.visit_location,
    ir.record_id,
    ir.auditor_score,
    ir.surveyor_score,
    ir.assigned_employee_id,
    ir.employee_name
FROM
    Incorrect_records ir
WHERE
    ir.employee_name IN (SELECT employee_name FROM suspect_list);




Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

71 rows affected.

+----------------+-----------+---------------+----------------+----------------------+---------------+
| visit_location | record_id | auditor_score | surveyor_score | assigned_employee_id | employee_name |
+----------------+-----------+---------------+----------------+----------------------+---------------+
|   AmRu14842    |   47055   |       2       |       10       |          1           |  Bello Azibo  |
|   HaYa21395    |   28729   |       2       |       10       |          1           |  Bello Azibo  |
|   AkRu05741    |   19961   |       3       |       10       |          1           |  Bello Azibo  |
|   AmRu14887    |   31028   |       9       |       10       |          1           |  Bello Azibo  |
|   KiMr24919    |   23328   |       1       |       10       |          1           |  Bello Azibo  |
|   KiRu29639    |   45008   |       1       |       10       |          1           |  Bello Azibo  |
|   KiAm22092    |    8186   |       9       |       10       |          1           |  Bello Azibo  |
|   AkLu02500    |   24797   |       3       |       10       |          1           |  Bello Azibo  |
|   AkRu03848    |   43412   |       3       |       10       |          1           |  Bello Azibo  |
|   AkRu04508    |   37357   |       1       |       10       |          1           |  Bello Azibo  |
+----------------+-----------+---------------+----------------+----------------------+---------------+
Truncated to displaylimit of 10.

In [199]:
%%sql
SELECT 
    * 
FROM 
    auditor_report;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1620 rows affected.

+-------------+----------------------+-------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+
| location_id | type_of_water_source | true_water_source_score |                                                                 statements                                                                |
+-------------+----------------------+-------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+
|  SoRu34980  |         well         |            1            |          Residents admired the official's commitment to enhancing urban life, praising their cooperative and inclusive approach.          |
|  AkRu08112  |         well         |            3            | Villagers spoke highly of the official's dedication and genuine interest in their lives, fostering a sense of belonging and appreciation. |
|  AkLu02044  |        river         |            0            |         Villagers were touched by the official's interactions, noting their humility, strong work ethic, and respectful attitude.         |
|  AkHa00421  |         well         |            3            |   Villagers were moved by the official's visit, praising their hard work, humility, and the profound sense of connection they fostered.   |
|  SoRu35221  |        river         |            0            |                 A photographer's lens captures the queue, though his own struggle for water is a hidden part of the story.                |
|  HaAm16170  |         well         |            1            |               With an open heart, the official created an atmosphere of unity and familial camaraderie among the villagers.               |
|  AkRu04812  |         well         |            3            |     The official's presence left an indelible mark, reflecting their humility, dedication, and the genuine connections they nurtured.     |
|  AkRu08304  |         well         |            3            |         The official's interactions resonated deeply with the villagers, leaving a lasting impression of respect and camaraderie.         |
|  AkRu05107  |         well         |            2            | Villagers spoke highly of the official's dedication and genuine interest in their lives, fostering a sense of belonging and appreciation. |
|  AkRu05215  |         well         |            3            |                  Villagers admired the official's visit for its respectful interactions, hard work, and genuine concern.                  |
+-------------+----------------------+-------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+
Truncated to displaylimit of 10.

### Step 1: Update the Incorrect_records View

In [201]:
%%sql
CREATE OR REPLACE VIEW Incorrect_records AS (
    SELECT 
        vi.location_id AS visit_location,
        vi.record_id,
        ar.true_water_source_score AS auditor_score,
        wq.subjective_quality_score AS surveyor_score,
        vi.assigned_employee_id,
        e.employee_name,
        ar.statements -- Include the statements column
    FROM 
        auditor_report ar
    INNER JOIN
        visits vi
    ON 
        ar.location_id = vi.location_id
    AND
        vi.visit_count = 1
    JOIN
        water_quality wq
    ON 
        vi.record_id = wq.record_id 
    INNER JOIN
        employee e
    ON 
        vi.assigned_employee_id = e.assigned_employee_id
    WHERE
        vi.visit_count = 1 AND
        ar.true_water_source_score != wq.subjective_quality_score
);


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

++
||
++
++

### Step 2: Query to Retrieve Records for Suspect Employees

In [207]:
%%sql
WITH error_count AS (  
    SELECT
        employee_name,
        COUNT(*) AS number_of_mistakes
    FROM
        Incorrect_records
    GROUP BY
        employee_name
    ORDER BY
        number_of_mistakes DESC
),

average_mistakes AS( 
    SELECT 
        AVG(number_of_mistakes) AS avg_mistakes
    FROM 
        error_count
),

suspect_list AS (
    SELECT
        ec.employee_name
    FROM
        error_count ec
    JOIN
        average_mistakes am
    ON
        ec.number_of_mistakes > am.avg_mistakes
)

-- Retrieve records from Incorrect_records where employee_name is in the suspect_list
SELECbT
    ir.employee_name,
    ir.visit_location,  -- Use visit_location instead of location_id
    ir.statements
FROM
    Incorrect_records ir
WHERE
    ir.employee_name IN (SELECT employee_name FROM suspect_list);



Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

71 rows affected.

+---------------+----------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| employee_name | visit_location |                                                                                         statements                                                                                        |
+---------------+----------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|  Bello Azibo  |   AmRu14842    |                                    An elderly woman's weary eyes reflect the toll of daily hours spent in the queue, her thoughts on family back home.                                    |
|  Bello Azibo  |   HaYa21395    |                                       A university student's studies suffer from hours in line, dreams clouded by the daily reality of dirty water.                                       |
|  Bello Azibo  |   AkRu05741    |          An air of mistrust surrounded the official, as villagers spoke of laziness and hints of corruption. The mention of cash passing discreetly only deepened their concerns.         |
|  Bello Azibo  |   AmRu14887    |      Villagers expressed their discomfort with an official who displayed a haughty demeanor and negligence. The mention of cash transactions deepened their growing sense of unease.      |
|  Bello Azibo  |   KiMr24919    |      Suspicion and unease colored the villagers' accounts of an official's haughty behavior and potential corruption. The mention of cash changing hands added to their apprehension.     |
|  Bello Azibo  |   KiRu29639    |     An unsettling atmosphere prevailed as villagers shared stories of an official's arrogance and perceived corruption. The mention of cash exchanges only intensified their concerns.    |
|  Bello Azibo  |   KiAm22092    |                                       A mother-to-be's gentle waiting in line is filled with hope and concern, the water's quality a looming worry.                                       |
|  Bello Azibo  |   AkLu02500    |                                           A tour guide's enthusiasm is dampened by the water situation, his own wait in line a daily challenge.                                           |
|  Bello Azibo  |   AkRu03848    |                                     A local tailor's creativity is stifled by frustration, his work impacted by the corruption that taints the water.                                     |
|  Bello Azibo  |   AkRu04508    | An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts. |
+---------------+----------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
Truncated to displaylimit of 10.

In [263]:
%%sql
SELECT
    ir.employee_name,
    ir.visit_location,
    ir.statements
FROM
    Incorrect_records ir
WHERE

    ir.statements LIKE '%Suspicion colored%';


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

+----------------+----------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| employee_name  | visit_location |                                                                            statements                                                                            |
+----------------+----------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Lalitha Kaburi |   KiRu29329    | Suspicion colored villagers' descriptions of an official's aloof demeanor and apparent laziness. The reference to cash transactions cast doubt on their motives. |
+----------------+----------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+

In [212]:
%%sql
-- Filter records where statements mention 'cash'
SELECT
    ir.employee_name,
    ir.visit_location,
    ir.statements
FROM
    Incorrect_records ir
WHERE
    ir.statements LIKE '%cash%';


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

19 rows affected.

+----------------+----------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| employee_name  | visit_location |                                                                                              statements                                                                                              |
+----------------+----------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Zuriel Matembo |   SoRu38331    |      An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts.       |
| Malachi Mavuso |   AmAm09607    |            Villagers spoke of an unsettling encounter with an official who appeared dismissive and detached. The reference to cash transactions added to their growing sense of distrust.            |
|  Bello Azibo   |   KiIs23853    |            Villagers' wary accounts of an official's arrogance and detachment from their concerns raised suspicions. The mention of cash changing hands further tainted their perception.            |
|  Bello Azibo   |   HaSe21323    |            Villagers spoke of an unsettling encounter with an official who appeared dismissive and detached. The reference to cash transactions added to their growing sense of distrust.            |
| Zuriel Matembo |   AkRu05880    |               Villagers' wary accounts of an official's arrogance and detachment from their concerns raised suspicions. The allusion to cash changing hands deepened their skepticism.               |
|  Bello Azibo   |   KiRu27065    |           Villagers expressed their discomfort with an official who displayed a haughty demeanor and negligence. The mention of cash transactions deepened their growing sense of unease.            |
| Malachi Mavuso |   KiRu25347    |              Villagers expressed their discontent with an official who appeared dismissive and neglectful. The mention of cash changing hands added to their growing sense of distrust.              |
| Zuriel Matembo |   SoIl32575    | Villagers recounted unsettling encounters with an official known for their arrogance and avoidance of responsibilities. The mention of cash changing hands added to their apprehension and distrust. |
|  Bello Azibo   |   AkRu04508    |      An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts.       |
| Lalitha Kaburi |   AkRu07310    |                Villagers spoke of their unsettling encounters with an official who seemed indifferent and uninterested, hinting at potential improprieties involving cash exchanges.                 |
+----------------+----------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
Truncated to displaylimit of 10.

In [226]:
%%sql
WITH error_count AS (  
    SELECT
        employee_name,
        COUNT(*) AS number_of_mistakes
    FROM
        Incorrect_records
    GROUP BY
        employee_name
),

average_mistakes AS( 
    SELECT 
        AVG(number_of_mistakes) AS avg_mistakes
    FROM 
        error_count
),

suspect_list AS (
    SELECT
        ec.employee_name
    FROM
        error_count ec
    JOIN
        average_mistakes am
    ON
        ec.number_of_mistakes > am.avg_mistakes
)

-- Find employees mentioning 'cash' but not in the suspect_list
SELECT
    ir.employee_name,
     
   COUNT(ir.statements) AS Number_of_Statements
FROM
    Incorrect_records ir
WHERE
    ir.statements LIKE '%cash%'
    AND ir.employee_name IN (SELECT employee_name FROM suspect_list)
GROUP BY ir.employee_name
;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

4 rows affected.

+----------------+----------------------+
| employee_name  | Number_of_Statements |
+----------------+----------------------+
|  Bello Azibo   |          8           |
| Zuriel Matembo |          5           |
| Malachi Mavuso |          4           |
| Lalitha Kaburi |          2           |
+----------------+----------------------+

## Conculsion 

So we can sum up the evidence we have for Zuriel Matembo, Malachi Mavuso, Bello Azibo and Lalitha Kaburi:
1. They all made more mistakes than their peers on average.
2. They all have incriminating statements made against them, and only them.
Keep in mind, that this is not decisive proof, but it is concerning enough that we should flag it. Pres. Naledi has worked hard to stamp out
corruption, so she would urge us to report this.
 
I am a bit shocked to be honest! After all our teams set out to do, it is hard for me to uncover this. I'll let Pres. Naledi know what we found out.

In [229]:
%%sql
WITH Incorrect_records AS (
    SELECT
        auditorRep.location_id,
        visitsTbl.record_id,
        Empl_Table.employee_name,
        auditorRep.true_water_source_score AS auditor_score,
        wq.subjective_quality_score AS employee_score,
        auditorRep.statements AS statements,
        wp.pollutant_ppm,
        wp.biological
    FROM auditor_report AS auditorRep
    JOIN visits AS visitsTbl
        ON auditorRep.location_id = visitsTbl.location_id
    JOIN water_quality AS wq
        ON visitsTbl.record_id = wq.record_id
    JOIN employee as Empl_Table
        ON Empl_Table.assigned_employee_id = visitsTbl.assigned_employee_id
    JOIN well_pollution AS wp
        ON visitsTbl.source_id = wp.source_id
    WHERE visitsTbl.visit_count = 1 
        AND auditorRep.true_water_source_score != wq.subjective_quality_score
)
SELECT
    employee_name,
    count(employee_name)
FROM Incorrect_records
GROUP BY employee_name;


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

13 rows affected.

+----------------+----------------------+
| employee_name  | count(employee_name) |
+----------------+----------------------+
|   Rudo Imani   |          2           |
|  Bello Azibo   |          16          |
| Jengo Tumaini  |          2           |
|  Makena Thabo  |          3           |
| Lalitha Kaburi |          2           |
| Zuriel Matembo |          7           |
| Malachi Mavuso |          9           |
|  Enitan Zuri   |          2           |
| Thandiwe Kito  |          1           |
| Yewande Ebele  |          1           |
+----------------+----------------------+
Truncated to displaylimit of 10.

In [259]:
%%sql
SELECT
auditorRep.location_id,
visitsTbl.record_id,
auditorRep.true_water_source_score AS auditor_score,
wq.subjective_quality_score AS employee_score,
wq.subjective_quality_score - auditorRep.true_water_source_score  AS score_diff
FROM auditor_report AS auditorRep
JOIN visits AS visitsTbl
ON auditorRep.location_id = visitsTbl.location_id
JOIN water_quality AS wq
ON visitsTbl.record_id = wq.record_id
WHERE (wq.subjective_quality_score - auditorRep.true_water_source_score) > 9;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

10 rows affected.

+-------------+-----------+---------------+----------------+------------+
| location_id | record_id | auditor_score | employee_score | score_diff |
+-------------+-----------+---------------+----------------+------------+
|  SoBa31691  |    3267   |       0       |       10       |     10     |
|  AkHa00363  |   25387   |       0       |       10       |     10     |
|  SoKo33094  |   16159   |       0       |       10       |     10     |
|  KiRu27364  |   38102   |       0       |       10       |     10     |
|  KiRu27065  |   29772   |       0       |       10       |     10     |
|  KiRu29147  |    3179   |       0       |       10       |     10     |
|  SoIl32770  |    7548   |       0       |       10       |     10     |
|  KiRu29329  |   11962   |       0       |       10       |     10     |
|  SoRu39544  |   17929   |       0       |       10       |     10     |
|  SoRu38535  |   44282   |       0       |       10       |     10     |
+-------------+-----------+---------------+----------------+------------+
Truncated to displaylimit of 10.